app/src/main/java/org/thoughtcrime/securesms/backup/FullBackupExporter.java

```diff
@@ -0,0 +1,611 @@
+              statementBuilder.addParameters(BackupProtos.SqlStatement.SqlParameter.newBuilder().setStringParamter(cursor.getString(i)));
+            } else if (cursor.getType(i) == Cursor.FIELD_TYPE_FLOAT) {
+              statementBuilder.addParameters(BackupProtos.SqlStatement.SqlParameter.newBuilder().setDoubleParameter(cursor.getDouble(i)));
+            } else if (cursor.getType(i) == Cursor.FIELD_TYPE_INTEGER) {
+              statementBuilder.addParameters(BackupProtos.SqlStatement.SqlParameter.newBuilder().setIntegerParameter(cursor.getLong(i)));
+            } else if (cursor.getType(i) == Cursor.FIELD_TYPE_BLOB) {
+              statementBuilder.addParameters(BackupProtos.SqlStatement.SqlParameter.newBuilder().setBlobParameter(ByteString.copyFrom(cursor.getBlob(i))));
+            } else if (cursor.getType(i) == Cursor.FIELD_TYPE_NULL) {
+              statementBuilder.addParameters(BackupProtos.SqlStatement.SqlParameter.newBuilder().setNullparameter(true));
+            } else {
+              throw new AssertionError("unknown type?"  + cursor.getType(i));
+            }
+
+            if (i < cursor.getColumnCount()-1) {
+              statement.append(',');
+            }
+          }
+
+          statement.append(')');
+
+          EventBus.getDefault().post(new BackupEvent(BackupEvent.Type.PROGRESS, ++count));
+          outputStream.write(statementBuilder.setStatement(statement.toString()).build());
+
+          if (postProcess != null) {
+            count = postProcess.postProcess(cursor, count);
+          }
+        }
+      }
+    }
+
+    return count;
+  }
+
+  private static int exportAttachment(@NonNull AttachmentSecret attachmentSecret, @NonNull Cursor cursor, @NonNull BackupFrameOutputStream outputStream, int count) {
+    try {
+      long rowId    = cursor.getLong(cursor.getColumnIndexOrThrow(AttachmentDatabase.ROW_ID));
+      long uniqueId = cursor.getLong(cursor.getColumnIndexOrThrow(AttachmentDatabase.UNIQUE_ID));
+      long size     = cursor.getLong(cursor.getColumnIndexOrThrow(AttachmentDatabase.SIZE));
+
+      String data   = cursor.getString(cursor.getColumnIndexOrThrow(AttachmentDatabase.DATA));
+      byte[] random = cursor.getBlob(cursor.getColumnIndexOrThrow(AttachmentDatabase.DATA_RANDOM));
+
+      if (!TextUtils.isEmpty(data)) {
+        long fileLength = new File(data).length();
+        long dbLength   = size;
+
+        if (size <= 0 || fileLength != dbLength) {
+          size = calculateVeryOldStreamLength(attachmentSecret, random, data);
+          Log.w(TAG, "Needed size calculation! Manual: " + size + " File: " + fileLength + "  DB: " + dbLength + " ID: " + new AttachmentId(rowId, uniqueId));
+        }
+      }
+
+      if (!TextUtils.isEmpty(data) && size > 0) {
+        InputStream inputStream;
+
+        if (random != null && random.length == 32) inputStream = ModernDecryptingPartInputStream.createFor(attachmentSecret, random, new File(data), 0);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *inputStream*. Currently, there are execution paths that do not contain closure statements, for example, when BackupFrameOutputStream.write (in method write) throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).